lnrpc-ideal/rpc.proto

```diff
@@ -0,0 +1,809 @@
+syntax = "proto3";
+
+import "google/api/annotations.proto";
+
+package lnrpc;
+
+option go_package = "github.com/lightningnetwork/lnd/lnrpc";
+
+service Lightning {
+    /** lncli: `listinvoices`
+    ListInvoices returns a list of all the invoices currently stored within the
+    database. Any active debug invoices are ignored. It has full support for
+    paginated responses, allowing users to query for specific invoices through
+    their add_index. This can be done by using either the first_index_offset or
+    last_index_offset fields included in the response as the index_offset of the
+    next request. By default, the first 100 invoices created will be returned.
+    Backwards pagination is also supported through the Reversed flag.
+    */
+    rpc ListInvoices(ListInvoiceRequest) returns (ListInvoiceResponse) {
+      option (google.api.http) = {
+        get : "/v1/invoices"
+      };
+    }
+  
+    /** lncli: `lookupinvoice`
+    LookupInvoice attempts to look up an invoice according to its payment hash.
+    The passed payment hash *must* be exactly 32 bytes, if not, an error is
+    returned.
+    */
+    rpc LookupInvoice(PaymentHash) returns (Invoice) {
+      option (google.api.http) = {
+        get : "/v1/invoice/{r_hash_str}"
+      };
+    }
+  
+    /**
+    SubscribeInvoices returns a uni-directional stream (server -> client) for
+    notifying the client of newly added/settled invoices. The caller can
+    optionally specify the add_index and/or the settle_index. If the add_index
+    is specified, then we'll first start by sending add invoice events for all
+    invoices with an add_index greater than the specified value. If the
+    settle_index is specified, the next, we'll send out all settle events for
+    invoices with a settle_index greater than the specified value. One or both
+    of these fields can be set. If no fields are set, then we'll only send out
+    the latest add/settle events.
+    */
+    rpc SubscribeInvoices(InvoiceSubscription) returns (stream Invoice) {
+      option (google.api.http) = {
+        get : "/v1/invoices/subscribe"
+      };
+    }
+  
+    /** lncli: `listpayments`
+    ListPayments returns a list of all outgoing payments.
+    */
+    rpc ListPayments(ListPaymentsRequest) returns (ListPaymentsResponse) {
```

Comment:
Suppose you have a wallet that needs to show an (updating) overview of all payments (finished and in-flight), how would that be implemented efficiently? This call can be polled, but the list may be long. A payment that is initiated a long time ago is on a high page number.

Reply:
Maybe this should be the guiding question: is it possible to create an efficiently updating (payment or invoice) monitor?

Reply:
> needs to show an (updating) overview of all payments (finished and in-flight)

Cutting this up to 3 parts:

New payments:
- Pretty simple, can expect the existing payment to be added to the list by the dev once they've sent it

Existing in flight payments:
- We want to get all existing in flight payments, regardless of when they were created. Timestamp probably won't work here, because we could have paid a hodl invoice which was last accepted a week ago. This would be where something like `TrackAllPayments` may come in? You want a subscription to payment events for every incomplete payment, regardless of age. 

Complete payments:
- Probably want the most recent x payments, can use listpayments as is in reverse, and just get a single page of results.

Reply:
What could be a relatively easy change to `ListPayments` is return a `stream lnrpc.Payment`. Then first stream all the payments as indicated by the request parameters and from there on, keep streaming updates on that same set of payments. For a wallet that show payments in pages, this would avoid polling.

Reply:
A different usage mode could be to not restrict the number of returned results (`max_payments` set to zero), but set a new parameter `min_last_update` that only streams payments and payment updates with a `last_update` timestamp of at least `min_last_update`. This could be used by applications that have their own persistent state and need to 'catch up' since a certain point.

Reply:
Maybe there are also use cases where it is beneficial to combine both filters